x-pack/plugins/security_solution/server/lib/detection_engine/notifications/legacy_saved_object_references/legacy_inject_rule_id_references.ts

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Logger, SavedObjectReference } from 'src/core/server';
+// eslint-disable-next-line no-restricted-imports
+import { LegacyRulesNotificationParams } from '../legacy_types';
+
+/**
+ * This injects any legacy "id"'s from saved object reference and returns the "ruleAlertId" using the saved object reference. If for
+ * some reason it is missing on saved object reference, we log an error about it and then take the last known good value from the "ruleId"
+ *
+ * @deprecated Once we are confident all rules relying on side-car actions SO's have been migrated to SO references we should remove this function
+ * @param logger The kibana injected logger
+ * @param ruleAlertId The alert id to merge the saved object reference from.
+ * @param savedObjectReferences The saved object references which should contain a "ruleAlertId"
+ * @returns The "ruleAlertId" with the saved object reference replacing any value in the saved object's id.
+ */
+export const legacyInjectRuleIdReferences = ({
+  logger,
+  ruleAlertId,
+  savedObjectReferences,
+}: {
+  logger: Logger;
+  ruleAlertId: LegacyRulesNotificationParams['ruleAlertId'];
+  savedObjectReferences: SavedObjectReference[];
+}): LegacyRulesNotificationParams['ruleAlertId'] => {
+  const referenceFound = savedObjectReferences.find((reference) => {
+    return reference.name === 'alert_0';
+  });
+  if (referenceFound) {
+    if (referenceFound.id !== ruleAlertId) {
+      // This condition should not be reached but we log an error if we encounter it to help if we migrations
+      // did not run correctly or we create a regression in the future.
+      logger.error(
+        [
+          'The id of the "saved object reference id": ',
+          referenceFound.id,
+          ' is not the same as the "saved object id": ',
+          ruleAlertId,
+          '. Preferring and using the "saved object reference id" instead of the "saved object id"',
+        ].join('')
+      );
+    }
+    return referenceFound.id;
+  } else {
+    logger.error(
+      [
+        'The saved object reference was not found for the "ruleAlertId" when we were expecting to find it. ',
+        'Kibana migrations might not have run correctly or someone might have removed the saved object references manually. ',
+        'Returning the last known good "ruleAlertId" which might not work. "ruleAlertId" with its id being returned is: ',
```

Comment:
What's worst case here? 

1) last "ruleAlertId" is the expected one and all works fine
2) last "ruleAlertId" is not what was expected and user has a "broken" alert attached

Reply:
Those are both correct, but case 2 is the mostly likely thing to happen.

Worst case is that the alertId doesn't point to anything valid and they are broken. It's the same issue as above which is I cannot guarantee within Elastic referential integrity. I think most programmers and code just _assumes_ this case wouldn't happen where I am adding code to check and report it.

This really shouldn't happen easily.

Also the last known good `ruleAlertId` might not be there at all.